test/test_unary_ufuncs.py

```diff
@@ -0,0 +1,75 @@
+import torch
+
+from torch.testing._internal.common_utils import \
+    (TestCase, run_tests)
+from torch.testing._internal.common_methods_invocations import \
+    (unary_ufuncs)
+from torch.testing._internal.common_device_type import \
+    (instantiate_device_type_tests, ops, onlyOnCPUAndCUDA)
+
+
+# Returns a tensor of the requested shape, dtype, and device
+# Initialization uses randint for non-float types and randn for float types.
+def _make_tensor(shape, dtype, device, fill_ones=False) -> torch.Tensor:
+    # Returns a tensor filled with ones
+    if fill_ones:
+        return torch.ones(*shape, dtype=_convert_t(dtype, device), device=device)
+
+    # Returns a tensor with random integer values
+    if not (dtype.is_floating_point or dtype.is_complex):
+        t = torch.randint(0, 10, shape, device=device)
+        if dtype != torch.uint8:
+            t = t - 5  # generate negative values also
+        return t.to(_convert_t(dtype, device))
+
+    # Populates the CPU tensor with floats representable as half/bfloat16
+    if dtype == torch.half and device == 'cpu':
+        return torch.randn(*shape, dtype=torch.float, device=device).half()
+    if dtype == torch.bfloat16 and device == 'cpu':
+        return torch.randn(*shape, dtype=torch.float, device=device).bfloat16()
+
+    # Default: returns a tensor with random float values
+    return torch.randn(shape, dtype=dtype, device=device)
+
+# Tests for unary "universal functions (ufuncs)" that accept a single
+# tensor and have common properties like:
+#   - they are elementwise functions
+#   - the input shape is the output shape
+#   - they typically have method and inplace variants
+#   - they typically support the out kwarg
+#   - they typically have NumPy or SciPy references
+
+# See NumPy's universal function documentation
+# (https://numpy.org/doc/1.18/reference/ufuncs.html) for more details
+# about the concept of ufuncs.
+class TestUnaryUfuncs(TestCase):
+    exact_dtype = True
+
+    # Verifies that the unary ufuncs have their supported dtypes
+    #   registered correctly by testing that each unlisted dtype
+    #   throws a runtime error
+    @onlyOnCPUAndCUDA
+    @ops(unary_ufuncs, unsupported_dtypes_only=True)
+    def test_unsupported_dtypes(self, device, dtype, op_meta):
+        t = torch.empty(1, device=device, dtype=dtype)
+        with self.assertRaises(RuntimeError):
+            op = op_meta.getOp()
+            op(t)
+
+    @ops(unary_ufuncs)
+    def test_non_contig(self, device, dtype, op_meta):
```

Comment:
Perhaps I am taking this example a little too literally, but if I were trying to uniformly test that noncontiguous inputs are handled correctly across all operators inside of PyTorch, this isn't the way I would go about doing it. The big problem is that there is a lot of commonality for handling of non-contiguous inputs beyond just unary ufuncs, and so if I divide the tests by classification like `unary_ufuncs`, I'd end up having to substantially replicating the logic in this test for every single "class" of ops I'd test. Whereas what I'd hope for is to abstract things even more, where I just have an abstract notion of "Here is an op I can call, here are some valid input sizes" and then in my test I'd arrange for those sizes to be contiguous or not contiguous. I'd like to write a *single* test for *all* operators, not just unary ufuncs.

By the way, the duplication wouldn't be so bad if the test would be as simple as it's written here. But there's quite a lot of complexity in how we handle noncontiguous inputs. With operators that just call `contiguous()` and do their thing, a simple test like this will do fine; but if you are, for example, trying to indirectly test TensorIterator, there'd be a lot more cases I'd be looking to cover. And even treating the test as a black box, there are different flavors of noncontiguity (like broadcasting inputs, or inputs that have memory overlap)

Reply:
This complaint applies, though to a lesser degree, to `test_unsupported_dtypes` too: unsupported dtype test is also a test that I'd expect to generalize to any operator, not just unary ufuncs. And a generalized test that can handle multiple arguments, would also work for unary ufuncs.

Maybe there is an implication that it is too hard to write "super-metaprogrammed" tests in this way. If so, we should explicitly call it out and discuss this.

Reply:
I agree this is interesting to look at in the future, and it's great you see the possibilities for this work. What you're describing is entirely doable, and I think as more OpMetas and more metadata is added to ops it will be straightforward to do, if you want to. This is only a start, and it only copies one existing test as an example.